addons/mail/models/res_users.py

```diff
@@ -137,6 +137,28 @@ def _get_portal_access_update_body(self, access_granted):
             return '%s (%s)' % (body, self.partner_id.email)
         return body
 
+    @api.model
+    def _deactivate_portal_user(self, **post):
+        """Blacklist the email of the user after deleting it.
+
+        Log a note on the related partner so we know why it's archived.
+        """
+        for user in self:
+            user.partner_id._message_log(body=_(
+                'This partner was archived because the related user "%s" has requested to deactivate his account.',
+                user.name))
+
+        if post.get('blacklist'):
+            emails_to_blacklist = filter(bool, [tools.email_normalize(user.email) for user in self])
+        else:
+            emails_to_blacklist = []
+
+        super(Users, self)._deactivate_portal_user(**post)
+
+        for email in emails_to_blacklist:
+            blacklist = self.env['mail.blacklist']._add(email)
+            blacklist._message_log(body=_('The email has been blacklisted because its related portal user has deactivated his account.'))
```

Comment:
Maybe something more complete like "This emails is blacklisted because %(user_name)s (#%(user_id)s) asked to deactivate %(portal_user_name)s account and blacklist their emails" ?